pybop/samplers/annealed_importance.py

```diff
@@ -0,0 +1,138 @@
+        return self._chains
+
+    @chains.setter
+    def chains(self, value: int) -> None:
+        """Sets the total number of iterations."""
+        if not isinstance(value, (int, np.integer)):
+            raise TypeError("iterations must be an integer")
+        if value <= 0:
+            raise ValueError("iterations must be positive")
+        self._chains = int(value)
+
+    @property
+    def num_beta(self) -> int:
+        """Returns the number of beta points"""
+        return self._num_beta
+
+    def set_num_beta(self, num_beta: int) -> None:
+        """Sets the number of beta point values."""
+        if not isinstance(num_beta, (int, np.integer)):
+            raise TypeError("num_beta must be an integer")
+        if num_beta <= 1:
+            raise ValueError("num_beta must be greater than 1")
+        self._num_beta = num_beta
+        self._beta = np.linspace(0, 1, num_beta)
+
+    def transition_distribution(self, x, j):
+        """
+        Transition distribution for each beta value [j] - Eqn 3.
+        """
+        return (1.0 - self._beta[j]) * self._log_prior(x) + self._beta[
+            j
+        ] * self._log_likelihood(x)
+
+    def run(self) -> tuple[float, float, float]:
+        """
+        Run the annealed importance sampling algorithm.
+
+        Returns:
+            Tuple containing (mean, median, std, variance) of the log weights
+
+        Raises:
+            ValueError: If starting position has non-finite log-likelihood
+        """
+        log_w = np.zeros(self._chains)
+        I = np.zeros(self._chains)
+        samples = np.zeros(self._num_beta)
+
+        for i in range(self._chains):
+            current = self._log_prior.rvs()
+            if not np.isfinite(self._log_likelihood(current)):
+                raise ValueError("Starting position has non-finite log-likelihood.")
+
+            current_f = self._log_prior(current)
+
+            log_density_current = np.zeros(self._num_beta)
+            log_density_current[0] = current_f
+            log_density_previous = np.zeros(self._num_beta)
+            log_density_previous[0] = current_f
+
+            # Main sampling loop
+            for j in range(1, self._num_beta):
+                # Compute jth transition with current sample
+                log_density_current[j] = self.transition_distribution(current, j)
+
+                # Calculate the previous transition with current sample
+                log_density_previous[j] = self.transition_distribution(current, j - 1)
+
+                # Generate new sample from current (eqn.4)
+                proposed = np.random.multivariate_normal(current, self._cov0)
+
+                # Evaluate proposed sample
+                if np.isfinite(self._log_likelihood(proposed)):
+                    proposed_f = self.transition_distribution(proposed, j)
+
+                    # Metropolis acceptance
+                    acceptance_log_prob = proposed_f - current_f
```

Comment:
for the gaussian test case they use in the paper, a much more complicated transition `T_j` is used, a sequence of 3 metropolis tests repeated 5-10 times. I'm not sure if all that is neccessary however